lib/internal/http2/core.js

```diff
@@ -983,22 +986,79 @@ function emitClose(self, error) {
   self.emit('close');
 }
 
-function finishSessionDestroy(session, error) {
+function cleanupSession(session) {
   const socket = session[kSocket];
-  if (!socket.destroyed)
-    socket.destroy(error);
-
   session[kProxySocket] = undefined;
   session[kSocket] = undefined;
   session[kHandle] = undefined;
   session[kNativeFields] = new Uint8Array(kSessionUint8FieldCount);
-  socket[kSession] = undefined;
-  socket[kServer] = undefined;
+  if (socket) {
+    socket[kSession] = undefined;
+    socket[kServer] = undefined;
+  }
+}
+
+function finishSessionDestroy(session, error) {
+  debugSessionObj(session, 'finishSessionDestroy');
+
+  const handle = session[kHandle];
+  const socket = session[kSocket];
+  if (handle) handle.ondone = null;
+  cleanupSession(session);
+
+  if (socket && !socket.destroyed)
+    socket.destroy(error);
 
   // Finally, emit the close and error events (if necessary) on next tick.
   process.nextTick(emitClose, session, error);
 }
 
+function finishSessionClose(session) {
+  debugSessionObj(session, 'finishSessionClose');
+
+  const socket = session[kSocket];
+  const handle = session[kHandle];
+  if (handle) handle.ondone = null;
+  cleanupSession(session);
+
+  if (!socket.destroyed) {
+    socket.end(() => emitClose(session));
```

Comment:
Note if this is going to be backported to v12, `end(cb)` doesn't always invoke the callback.

Reply:
Also, the callback may or may not (depending on Node version) be invoked with an error. Might cause `emitClose` to be invoked twice? Once through here and once through error handler?

Reply:
This may be an issue. I think we can manually do `once('finish')`, `once('error')` for v12, would that be okay?

> Once through here and once through error handler?

Could you clarify what error handler? If you refer to `socketOnError` then that is basically noop after `cleanupSession()` as we remove `socket[kSession]`.

Also, this raises another question, previously any error from socket would be ignored in here, should we perhaps forward it to `emitClose` now?

Reply:
> I think we can manually do once('finish'), once('error') for v12, would that be okay?

Yes, I think so.

> Also, this raises another question, previously any error from socket would be ignored in here, should we perhaps forward it to emitClose now?

Yes, I think so since.

> oop after cleanupSession() as we remove socket[kSession]

Reply:
Actually, this might be a problem for v13 as well, not sure where/when the `end(cb)` fix was merged.

Reply:
https://github.com/nodejs/node/pull/29747